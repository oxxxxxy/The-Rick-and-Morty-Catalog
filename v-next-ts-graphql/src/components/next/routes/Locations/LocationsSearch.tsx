import { useState, useRef } from 'react';


import {
	API_LOCATIONS__PARAM__NAME,
	API_LOCATIONS__PARAM__DIMENSION,
	API_LOCATIONS__PARAM__TYPE,
	API_LOCATIONS__PARAM_LIST
} from '@tsCF/data';

import type { 
	QueryParamCompatible_Base,
	QPC_InputText
} from '@tsLF/forURLSP';

import { CustomFormHolder } from '@tsLF/pages';
import type { 
	ArgumentsFor_CustomFormHolder,
	ValueStore,
	objWithFnsForEachCFIDC__get_exitValue
} from '@tsLF/pages';


import InputText from '@/components/next/customForm/InputText';




export default function LocationsSearch(
	{
		init_cachedValues,
		update_values,
		get_exitValue
	}:{
		init_cachedValues: QueryParamCompatible_Base[];
		update_values: QueryParamCompatible_Base[];
		get_exitValue: (v: QueryParamCompatible_Base[]) => void;
	}
){
	const [isValid, setIsValid] = useState<boolean>(true);
	const [update_valuesJson, setUpdate_valuesJson] = useState<string>(JSON.stringify(update_values));
	const set_value = get_exitValue;
	const set_applyActivity = (v: boolean) => setIsValid(v);
	
	const [exitValueStore, setExitValueStore] = useState<ValueStore>({});
	const [entryValueStore, setEntryValueStore] = useState<ValueStore>({});

	const customFormHolder = useRef<CustomFormHolder>();
	const [objWithFnsForEachCFIDC__get_exitValue] = useState<objWithFnsForEachCFIDC__get_exitValue>(
		() => {
			const CFIDCList = API_LOCATIONS__PARAM_LIST;
			
			const args: ArgumentsFor_CustomFormHolder = {
				set_value,
				set_applyActivity,
			
				CFIDCList: CFIDCList
			};
			
			if(init_cachedValues){
				args.cachedQPCValues = init_cachedValues;
			} 
		
			customFormHolder.current = new CustomFormHolder(args);
			
			setExitValueStore(CustomFormHolder.makeValueStore(CFIDCList));
			setEntryValueStore(CustomFormHolder.makeValueStore(CFIDCList));

			if(update_values.length){
				updateUpdate_values();
			}

			const objWithFnsForEachCFIDC__get_exitValue: objWithFnsForEachCFIDC__get_exitValue = {};
			
			for(const el of CFIDCList){
				objWithFnsForEachCFIDC__get_exitValue[el.name] = (v: QPC_InputText) => {
					exitValueStore[el.name] = v;
					setExitValueStore(exitValueStore);
					// prosti menya, gospod'... no ya greshen...
					// @ts-ignore-next-line
					customFormHolder.current.recieveExitValueStore(exitValueStore);
				}
			}
			
			return objWithFnsForEachCFIDC__get_exitValue;
		}
	);


	if(!customFormHolder.current){
		throw new Error('OMG WE\'RE ALL GOING TO DIE!!!! Let\'s fuck in the asses, dudes.');
	}


	function updateUpdate_values(){
		const storeValue = CustomFormHolder.makeValueStore(API_LOCATIONS__PARAM_LIST);
		CustomFormHolder.setValuesToValueStore(storeValue, update_values);
		setEntryValueStore(storeValue);
		setUpdate_valuesJson(JSON.stringify(update_values));
	}

	if(update_valuesJson !== JSON.stringify(update_values)){
		updateUpdate_values();
	}


	const apply = () => {
		// prosti menya, gospod'... no ya greshen...
		// @ts-ignore-next-line
		customFormHolder.current.apply();
	}


	return (
<div className="margin-10 w-100 d-flex jc-center">
  <div
    className="search--width d-flex jc-space-between fd-column"
    title="locations"
  >

		<InputText
			get_exitValue = {
				objWithFnsForEachCFIDC__get_exitValue[
					API_LOCATIONS__PARAM__NAME
					.name
				]
			}
			entry_value = {
				entryValueStore[
					API_LOCATIONS__PARAM__NAME
					.name
				]
			}
			// prosti menya, gospod'... no ya greshen...
			// @ts-ignore-next-line
			init_instanceOfInputText = {
				customFormHolder.current.getInstanceOfCFItemFor(
					API_LOCATIONS__PARAM__NAME
				)
			}

			key={
						// API_LOCATIONS__PARAM__NAME
						// .name + 
				JSON.stringify(
					entryValueStore[
						API_LOCATIONS__PARAM__NAME
						.name
					]
				)
			}
		/>
		<InputText
			get_exitValue = {
				objWithFnsForEachCFIDC__get_exitValue[
					API_LOCATIONS__PARAM__TYPE
					.name
				]
			}
			entry_value = {
				entryValueStore[
					API_LOCATIONS__PARAM__TYPE
					.name
				]
			}
			// prosti menya, gospod'... no ya greshen...
			// @ts-ignore-next-line
			init_instanceOfInputText = {
				customFormHolder.current.getInstanceOfCFItemFor(
					API_LOCATIONS__PARAM__TYPE
				)
			}
			
			key={
						// API_LOCATIONS__PARAM__TYPE
						// .name +
				JSON.stringify(
					entryValueStore[
						API_LOCATIONS__PARAM__TYPE
						.name
					]
				)
			}
		/>
		<InputText
			get_exitValue = {
				objWithFnsForEachCFIDC__get_exitValue[
					API_LOCATIONS__PARAM__DIMENSION
					.name
				]
			}
			entry_value = {
				entryValueStore[
					API_LOCATIONS__PARAM__DIMENSION
					.name
				]
			}
			// prosti menya, gospod'... no ya greshen...
			// @ts-ignore-next-line
			init_instanceOfInputText = {
				customFormHolder.current.getInstanceOfCFItemFor(
					API_LOCATIONS__PARAM__DIMENSION
				)
			}

			key={
						// API_LOCATIONS__PARAM__DIMENSION
						// .name + 
				JSON.stringify(
					entryValueStore[
						API_LOCATIONS__PARAM__DIMENSION
						.name
					]
				)
			}
		/>


    <div className="filter-select-box d-flex jc-center ai-center">


    </div>

    <button
      className={`
				filter-button color--b6b6b6 bg-color--181a1b tt-uppercase 
				${ isValid ? 'button--has-some': 'button--empty'}
			`}

			disabled = {!isValid}

		  onClick={apply}
    >
      Apply
    </button>
  </div>
</div>
	);
}
