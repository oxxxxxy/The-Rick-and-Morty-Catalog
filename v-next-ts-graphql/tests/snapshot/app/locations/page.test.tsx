import { render } from '@testing-library/react';
import { expect, test } from 'vitest';
import Locations from '@/app/locations/page';

test('matches snapshot', () => {
  const { container } = render(<Locations />);
  expect(container).toMatchSnapshot();
});

