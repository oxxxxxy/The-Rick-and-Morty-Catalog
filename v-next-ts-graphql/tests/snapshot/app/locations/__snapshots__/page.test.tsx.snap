// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <div
    class="
				d-flex
				ai-center
				jc-center
				fd-column
				color--b6b6b6
				main--font-size
			"
  >
    <div
      class="
					d-flex
					jc-center
					margin-10
					w-100
				"
    >
      <div
        class="
						search--width
						d-flex
						ai-center
						jc-center
						fd-if-d-row-if-m-column
					"
      >
        <span
          class="
							search-word
						"
        >
          Search :
        </span>
        <div
          class="
							select-list-option-line
							d-flex
						"
        >
          <a
            class="
									select-list-option
									underline
									
								"
            href="/characters"
            title="Search {capitalizeWord(path.value)}"
          >
            characters
          </a>
          <a
            class="
									select-list-option
									underline
									selected-select-list-option
								"
            href="/locations"
            title="Search {capitalizeWord(path.value)}"
          >
            locations
          </a>
          <a
            class="
									select-list-option
									underline
									
								"
            href="/episodes"
            title="Search {capitalizeWord(path.value)}"
          >
            episodes
          </a>
        </div>
        <span
          class="
							search-word-dot
						"
        >
          .
        </span>
      </div>
    </div>
    <div
      class="margin-10 w-100 d-flex jc-center"
    >
      <div
        class="search--width d-flex jc-space-between fd-column"
        title="characters"
      >
        <div
          class="text-input-option d-flex w-100 fd-column"
        >
          <div
            class="fd-row d-flex w-100"
          >
            <div
              class="text-input-box"
            >
              <div
                class="text-input-underline bg-color--282828"
              />
              <input
                class="text-input bg-color--b6b6b6 color--282828"
                placeholder="Name (e.g. Earth (Replacement Dimension))"
                type="text"
                value=""
              />
            </div>
            <button
              class="
				  	ai-center jc-center text-input-clear fill--999999 bg-color--181a1b
						button--empty
				  "
              disabled=""
            >
              <svg
                class="text-input-clear-icon"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.83 5.146a.5.5 0 0 0 0 .708L7.975 8l-2.147 2.146a.5.5 0 0 0 .707.708l2.147-2.147 2.146 2.147a.5.5 0 0 0 .707-.708L9.39 8l2.146-2.146a.5.5 0 0 0-.707-.708L8.683 7.293 6.536 5.146a.5.5 0 0 0-.707 0z"
                />
                <path
                  d="M13.683 1a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-7.08a2 2 0 0 1-1.519-.698L.241 8.65a1 1 0 0 1 0-1.302L5.084 1.7A2 2 0 0 1 6.603 1zm-7.08 1a1 1 0 0 0-.76.35L1 8l4.844 5.65a1 1 0 0 0 .759.35h7.08a1 1 0 0 0 1-1V3a1 1 0 0 0-1-1z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="text-input-option d-flex w-100 fd-column"
        >
          <div
            class="fd-row d-flex w-100"
          >
            <div
              class="text-input-box"
            >
              <div
                class="text-input-underline bg-color--282828"
              />
              <input
                class="text-input bg-color--b6b6b6 color--282828"
                placeholder="Type (e.g. Planet)"
                type="text"
                value=""
              />
            </div>
            <button
              class="
				  	ai-center jc-center text-input-clear fill--999999 bg-color--181a1b
						button--empty
				  "
              disabled=""
            >
              <svg
                class="text-input-clear-icon"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.83 5.146a.5.5 0 0 0 0 .708L7.975 8l-2.147 2.146a.5.5 0 0 0 .707.708l2.147-2.147 2.146 2.147a.5.5 0 0 0 .707-.708L9.39 8l2.146-2.146a.5.5 0 0 0-.707-.708L8.683 7.293 6.536 5.146a.5.5 0 0 0-.707 0z"
                />
                <path
                  d="M13.683 1a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-7.08a2 2 0 0 1-1.519-.698L.241 8.65a1 1 0 0 1 0-1.302L5.084 1.7A2 2 0 0 1 6.603 1zm-7.08 1a1 1 0 0 0-.76.35L1 8l4.844 5.65a1 1 0 0 0 .759.35h7.08a1 1 0 0 0 1-1V3a1 1 0 0 0-1-1z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="text-input-option d-flex w-100 fd-column"
        >
          <div
            class="fd-row d-flex w-100"
          >
            <div
              class="text-input-box"
            >
              <div
                class="text-input-underline bg-color--282828"
              />
              <input
                class="text-input bg-color--b6b6b6 color--282828"
                placeholder="Dimension (e.g. Replacement Dimension)"
                type="text"
                value=""
              />
            </div>
            <button
              class="
				  	ai-center jc-center text-input-clear fill--999999 bg-color--181a1b
						button--empty
				  "
              disabled=""
            >
              <svg
                class="text-input-clear-icon"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.83 5.146a.5.5 0 0 0 0 .708L7.975 8l-2.147 2.146a.5.5 0 0 0 .707.708l2.147-2.147 2.146 2.147a.5.5 0 0 0 .707-.708L9.39 8l2.146-2.146a.5.5 0 0 0-.707-.708L8.683 7.293 6.536 5.146a.5.5 0 0 0-.707 0z"
                />
                <path
                  d="M13.683 1a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-7.08a2 2 0 0 1-1.519-.698L.241 8.65a1 1 0 0 1 0-1.302L5.084 1.7A2 2 0 0 1 6.603 1zm-7.08 1a1 1 0 0 0-.76.35L1 8l4.844 5.65a1 1 0 0 0 .759.35h7.08a1 1 0 0 0 1-1V3a1 1 0 0 0-1-1z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="filter-select-box d-flex jc-center ai-center"
        />
        <button
          class="
				filter-button color--b6b6b6 bg-color--181a1b tt-uppercase 
				button--has-some
			"
        >
          Apply
        </button>
      </div>
    </div>
  </div>
  <div
    class="
				main--font-size
				color--b6b6b6
				d-flex
				ai-center
				fd-column
			"
  >
    <div
      class="
					result-board
				"
    >
      <p>
        Loading...
      </p>
    </div>
  </div>
</div>
`;
